src/formElementPlugins/listBuilder/src/ApplicationView.tsx

```diff
@@ -0,0 +1,339 @@
+    if (anyErrorItems(currentInputResponses, inputFields)) {
+      setInputError(true)
+      return
+    }
+    // Add item
+    if (selectedListItem === null) {
+      setListItems([...listItems, currentInputResponses])
+    } else {
+      // Or update existing item
+      const newList = [...listItems]
+      newList[selectedListItem] = currentInputResponses
+      setListItems(newList)
+    }
+    setCurrentInputResponses(resetCurrentResponses(inputFields))
+    setOpen(false)
+    setSelectedListItem(null)
+  }
+
+  const editItem = async (index: number) => {
+    if (!isEditable) return
+    setSelectedListItem(index)
+    setCurrentInputResponses(listItems[index])
+    setOpen(true)
+  }
+
+  const deleteItem = async (index: number) => {
+    setListItems(listItems.filter((_, i) => i !== index))
+    setOpen(false)
+    setCurrentInputResponses(resetCurrentResponses(inputFields))
+  }
+
+  const listDisplayProps: ListLayoutProps = {
+    listItems,
+    displayFormat,
+    editItem,
+    deleteItem,
+    fieldTitles: inputFields.map((e: TemplateElement) => e.title),
+    codes: inputFields.map((e: TemplateElement) => e.code),
+    Markdown,
+    isEditable,
+  }
+
+  const DisplayComponent =
+    displayType === DisplayType.TABLE ? (
+      <ListTableLayout {...listDisplayProps} />
+    ) : (
+      <ListCardLayout {...listDisplayProps} />
+    )
+
+  return (
+    <>
+      <label>
+        <Markdown text={label} semanticComponent="noParagraph" />
+      </label>
+      <Markdown text={description} />
+      <Button
+        primary
+        content={createModalButtonText}
+        onClick={() => setOpen(true)}
+        disabled={!isEditable}
+      />
+      <Modal
+        size="tiny"
+        onClose={() => {
+          setOpen(false)
+          setSelectedListItem(null)
+          setCurrentInputResponses(resetCurrentResponses(inputFields))
+          setInputError(false)
+        }}
+        onOpen={() => setOpen(true)}
+        open={open}
+      >
+        <Segment>
+          <Form>
+            <Markdown text={modalText} />
+            {inputFields.map((field: TemplateElement, index: number) => {
+              const element = buildElement(field, index)
+              return (
+                <ApplicationViewWrapper
+                  key={`list-${element.code}`}
+                  element={element}
+                  isStrictPage={inputError}
+                  allResponses={allResponses}
+                  currentResponse={currentInputResponses[element.code].value}
+                  onSaveUpdateMethod={innerElementUpdate(element.code)}
+                  applicationData={applicationData}
+                />
+              )
+            })}
+            <Button
+              primary
+              content={selectedListItem !== null ? updateButtonText : addButtonText}
+              onClick={updateList}
+            />
+            {displayType === 'table' && selectedListItem !== null && (
```

Comment:
recently I've been liking the idea of keeping conditions in JSX free of 'logic'. As in extracting into constants and using constants (not suggesting a change, but just communicating my recent preference).
i.e. 

shouldRenderDeleteButton &&

Reply:
Sure, will bear it in mind.